File: client/vectors.ts
export type Vector = { x: number, y: number }
export type Direction = "North" | "South" | "East" | "West";
type V2 = Vector;

export function vector(x: number, y: number): Vector {
    return {x: Math.round(x), y: Math.round(y) };
}
export function equals(a: V2, b: V2): boolean { return a.x == b.x && a.y == b.y; }
export function add(a: V2, b: V2): V2 {
    return {
        x: a.x + b.x,
        y: a.y + b.y
    }
}
export function subtract(a: V2, b: V2): V2 {
    return {
        x: a.x - b.x,
        y: a.y - b.y
    }
}
export function multiply(a: V2, scalar: number) { return vector(a.x*scalar, a.y*scalar); }
export function zero(): V2 { return vector(0, 0); }
export function up(){ return vector(0, -1); }
export function down(){ return vector(0, 1); }
export function left(){ return vector(-1, 0); }
export function right(){ return vector(1, 0); }

export function dirToVec(direction: Direction){
    switch(direction){
        default: return zero();
        case "North": return up();
        case "South": return down();
        case "West": return left();
        case "East": return right();
    };
}

File: server/domain/vectors.py
from dataclasses import dataclass
from typing import Callable, List, Optional, Tuple


@dataclass(frozen=True)
class Vector:
    x: int
    y: int


def dir_to_vec(direction: str) -> Vector:
    """Convert a cardinal direction to a vector"""
    dir_vec_map = {
        # TODO - handle flipping the orientation to "Screen Space" on the
        # the client.  y=-1 for north is confusing
        "North": Vector(0, -1),
        "South": Vector(0, 1),
        "East": Vector(1, 0),
        "West": Vector(-1, 0),
    }
    return dir_vec_map.get(direction, Vector(0, 0))


def add(a: Vector, b: Vector) -> Vector:
    return Vector(a.x + b.x, a.y + b.y)


def subtract(a: Vector, b: Vector) -> Vector:
    return Vector(a.x - b.x, a.y - a.y)


def multiply(a: Vector, scalar: int):
    return Vector(a.x * scalar, a.y * scalar)


def raycast(start: Vector,
            direction: Vector,
            max_dist: int = 10,
            predicate: Callable[[Vector], bool] = lambda x: True
            ) -> Tuple[Optional[Vector], List[Vector]]:
    """
    Cast a ray starting from position `start` in Vector `direction`.
    Return a tuple (hit, positions) where `hit` is the first Vector position
    that fails the `predicate`, and positions is a list of positions (Vectors)
    to a max distance of `maxDist` that pass the predicate
    """
    ray = (add(start, multiply(direction, i + 1)) for i in range(max_dist))
    hit = None
    ray_path = []
    for e in ray:
        if predicate(e):
            ray_path.append(e)
        else:
            hit = e
            break
    return (hit, ray_path)
